File: scripts/dmzMBRAPluginFTPath.lua
local FTNameHandle = dmz.handle.new ("FT_Name")
local FTLinkHandle = dmz.handle.new ("FT_Link")
local FTPathHandle = dmz.handle.new ("FT_Node_Path_Name")
local iparis = iparis
local pairs = pairs

local function update_path (path, handle)
   local name = dmz.object.text (handle, FTNameHandle)
   if name then
      if path == "" then path = name
      else path = path .. "." .. name
      end
   end
   dmz.object.text (handle, FTPathHandle, path)
   local list = dmz.object.sub_links (handle, FTLinkHandle)
   if list then for _, sub in ipairs (list) do update_path (path, sub) end end
end

local function find_root (handle)
   local result = handle
   local list = dmz.object.super_links (handle, FTLinkHandle)
   if list then  result = find_root (list[1]) end
   return result
end

local function work (self)
   if self.list then
      local roots = {}
      for handle, _ in pairs (self.list) do
         roots[find_root (handle)] = true
      end
      for handle, _ in pairs (roots) do
         update_path ("", handle)
      end
      self.list = nil
   end
end

local function update_name (self, handle)
   if not self.list then self.list = {} end
   self.list[handle] = true
end

local function update_link (self, link, attr, super, sub)
   if not self.list then self.list = {} end
   self.list[super] = true
   self.list[sub] = true
end

local function start_plugin (self)
   self.timeSliceHandle = self.timeSlice:create (work, self, self.name)
   self.objObs:register (FTNameHandle, { update_object_text = update_name }, self)
   self.objObs:register (
      FTLinkHandle,
      { link_objects = update_link, unlink_objects = update_link },
      self)
end

local function stop_plugin (self)
   if self.timeSliceHandle then self.timeSlice:destroy (self.timeSliceHandle) end
end

function new (config, name)

   local self = {
      name = name,
      log = dmz.log.new ("lua." .. name),
      timeSlice = dmz.time_slice.new (),
      start_plugin = start_plugin,
      stop_plugin = stop_plugin,
      objObs = dmz.object_observer.new (),
   }

   self.log:info ("Creating plugin: " .. name)

   return self
end

File: scripts/dmzMBRAPluginFTPath.js
var dmz =
      { object: require("dmz/components/object")
      , defs: require("dmz/runtime/definitions")
      , time: require("dmz/runtime/time")
      }
   , FTNameHandle = dmz.defs.createNamedHandle("FT_Name")
   , FTLinkHandle = dmz.defs.createNamedHandle("FT_Link")
   , FTPathHandle = dmz.defs.createNamedHandle("FT_Node_Path_Name")
   , work
   , list = {}
   , updatePath
   , findRoot
   ;

updatePath = function (path, handle) {
   var name = dmz.object.text(handle, FTNameHandle)
     , subList
     , newPath = path
     ;

   if (name) {
      if (path == "") {
         newPath = name;
      }
      else {
         newPath += "." + name;
      }
   }
   dmz.object.text(handle, FTPathHandle, newPath);
   subList = dmz.object.subLinks(handle, FTLinkHandle);
   if (subList) {
      Object.keys(subList).forEach(function (key) {
         updatePath(newPath, subList[key]);
      });
   }
};

findRoot = function (handle) {
   var result = handle
     , superList = dmz.object.superLinks(handle, FTLinkHandle)
     ;
   if (superList) {
      result = findRoot(superList[0]);
   }
   return result;
};

work = dmz.time.setRepeatingTimer(self, function () {
   var roots
     , root
     ;
   if (list) {
      roots = {};
      Object.keys(list).forEach(function (key) {
         root = findRoot(list[key]);
         roots[root] = root;
      });
      Object.keys(roots).forEach(function (key) {
         updatePath("", roots[key]);
      });
      list = false;
   }
});

dmz.object.text.observe(self, FTNameHandle, function (handle) {
   if (!list) {
      list = {};
   }
   list[handle] = handle;
});

dmz.object.link.observe(self, FTLinkHandle, function (link, attr, Super, sub) {
   if (!list) {
      list = {};
   }
   list[Super] = Super;
   list[sub] = sub;
});

dmz.object.unlink.observe(self, FTLinkHandle, function (link, attr, Super, sub) {
   if (list) {
      if (list[Super]) {
         delete list[Super];
      }
      if (list[sub]) {
         delete list[sub];
      }
   }
});
